ovn-tester/ovn_context.py

```diff
@@ -32,27 +38,75 @@ def __exit__(self, type, value, traceback):
         ovn_stats.report(self.test_name, brief=self.brief_report)
         log.info(f'Exiting context {self.test_name}')
 
-    def __iter__(self):
-        return self
+    async def iteration_started(self, iteration):
+        '''Explicitly begin iteration 'n'. This is necessary when running
+        asynchronously since task starts and ends can overlap.'''
+        iteration.start()
+        log.info(f'Context {self.test_name}, Iteration {iteration.num}')
+        if self.test:
+            await self.test.exec_cmd(iteration.num, self.test_name)
 
-    def __next__(self):
-        now = time.perf_counter()
-        if self.iteration_start:
-            duration = now - self.iteration_start
-            ovn_stats.add(ITERATION_STAT_NAME, duration, failed=self.failed)
-            log.log(logging.WARNING if self.failed else logging.INFO,
-                    f'Context {self.test_name}, Iteration {self.iteration}, '
-                    f'Result: {"FAILURE" if self.failed else "SUCCESS"}')
-        self.failed = False
+    def iteration_completed(self, iteration):
+        ''' Explicitly end iteration 'n'. This is necessary when running
+        asynchronously since task starts and ends can overlap.'''
+        iteration.end()
+        duration = iteration.end_time - iteration.start_time
+        ovn_stats.add(ITERATION_STAT_NAME, duration, iteration.failed)
+        self.iterations = {task_name: it for task_name, it in
+                           self.iterations.items() if it != iteration}
         if self.test:
-            # exec external cmd
-            self.test.exec_cmd(self.iteration, self.test_name)
-        self.iteration_start = now
-        if self.iteration < self.max_iterations - 1:
-            self.iteration += 1
-            log.info(f'Context {self.test_name}, Iteration {self.iteration}')
-            return self.iteration
-        raise StopIteration
+            self.test.terminate_process(self.test_name)
+        log.log(logging.WARNING if iteration.failed else logging.INFO,
+                f'Context {self.test_name}, Iteration {iteration.num}, '
+                f'Result: {"FAILURE" if iteration.failed else "SUCCESS"}')
+
+    def create_task(self, coro, iteration=None):
+        '''Create a task to run in this context.'''
+        if iteration is None:
+            iteration = get_current_iteration()
+        task = asyncio.create_task(coro)
+        self.iterations[task.get_name()] = iteration
+        return task
+
+    async def qps_test(self, qps, coro, *args):
+        tasks = []
+        for i in range(self.max_iterations):
+            iteration = ContextIteration(i, self)
+            tasks.append(self.create_task(
+                self.qps_task(iteration, coro, *args), iteration)
+            )
+            if i % qps == 0:
+                await asyncio.sleep(1)
```

Comment:
This is cool!  Although, if I'm reading this properly, super nit: `if i != 0 and i % qps == 0:`

Reply:
I went a slightly different route:
```
if (i  + 1) % qps == 0:
    await asyncio.sleep(1)
```
The original version would sleep after iteration 0. Your suggestion fixes that problem, but it causes the first sleep to be performed after 21 iterations instead of 20.